tools/src/bin/build-blame.rs

```diff
@@ -0,0 +1,357 @@
+                    let blame_parent_entry = blame_parents[i].tree()?.get_path(&path)?;
+                    builder.insert(entry.name().unwrap(), blame_parent_entry.id(), entry.filemode())?;
+                    path.pop();
+                    continue 'outer;
+                }
+            }
+        }
+
+        match entry.kind() {
+            Some(ObjectType::Blob) => {
+                let blame_text = blame_for_path(
+                    file_movement,
+                    git_repo,
+                    commit,
+                    blame_repo,
+                    blame_parents,
+                    &path
+                )?;
+                builder.insert(entry.name().unwrap(), blame_repo.blob(&blame_text.as_bytes())?, entry.filemode())?;
+            }
+            Some(ObjectType::Commit) => {
+                // This is a submodule, just treat it as an empty dir. We could
+                // probably also skip over it entirely.
+                let entry_builder = blame_repo.treebuilder(None)?;
+                builder.insert(entry.name().unwrap(), entry_builder.write()?, entry.filemode())?;
+            }
+            Some(ObjectType::Tree) => {
+                let mut entry_builder = blame_repo.treebuilder(None)?;
+                build_blame_tree(
+                    &mut entry_builder,
+                    file_movement,
+                    git_repo,
+                    commit,
+                    blame_repo,
+                    blame_parents,
+                    path.clone(),
+                )?;
+                builder.insert(entry.name().unwrap(), entry_builder.write()?, entry.filemode())?;
+            }
+            _ => {
+                panic!("Unexpected entry kind {:?} found in tree for commit {:?} at path {:?}", entry.kind(), commit.id(), path);
+            }
+        };
+
+        path.pop();
+    }
+
+    Ok(())
+}
+
+fn main() {
+    env_logger::init();
+
+    let args: Vec<_> = env::args().collect();
+    let git_repo = Repository::open(&args[1]).unwrap();
+    let blame_repo = Repository::open(&args[2]).unwrap();
+    let use_cinnabar = env::var("CINNABAR").map(|v| v == "1").unwrap_or(false);
+    let blame_ref = env::var("BLAME_REF").ok().unwrap_or("HEAD".to_string());
+
+    info!("Reading existing blame map of ref {}...", blame_ref);
+    let mut blame_map = if let Ok(oid) = blame_repo.refname_to_id(&blame_ref) {
+        let (blame_map, _) = index_blame(&blame_repo, Some(oid));
+        blame_map
+    } else {
+        HashMap::new()
+    };
+
+    let mut refobj = blame_repo.find_reference(&blame_ref)
+        .ok()
+        .and_then(|r| r.resolve().ok());
+
+    let mut walk = git_repo.revwalk().unwrap();
+    walk.set_sorting(Sort::TOPOLOGICAL | Sort::REVERSE).unwrap();
+    walk.push(git_repo.refname_to_id(&blame_ref).unwrap()).unwrap();
+    let revs_to_process = walk
+        .map(|r| r.unwrap()) // walk produces Result<git2::Oid> so we unwrap to just the Oid
```

Comment:
You could also filter_map or such.

Reply:
This I left as-is because using `filter_map` would be cleanest if I did something like `r.ok().and_then(...)` but I don't really want to lose the `unwrap()` here and cause silent dropping of revs in case `libgit2` produces an err for whatever reason. Also it seems easier to read this way where the concerns are separated, although it might be less performant.